DIGNDB.App.SmitteStop.API/Services/RollingKeyPairGenerator.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Linq;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using Org.BouncyCastle.Asn1.X9;
+using Org.BouncyCastle.Crypto.Digests;
+using Org.BouncyCastle.Crypto.Generators;
+using Org.BouncyCastle.Crypto.Parameters;
+using Org.BouncyCastle.Math;
+
+namespace DIGNDB.App.SmitteStop.API.Services
+{
+    public class RollingKeyPairGenerator
+    {
+        private readonly byte[] _masterKeyBytes;
+
+        public X9ECParameters ECParameters { get; private set; }
+
+        public RollingKeyPairGenerator(byte[] masterKeyBytes, X9ECParameters ecParameters)
+        {
+            _masterKeyBytes = masterKeyBytes;
+            ECParameters = ecParameters;
+        }
+
+        public RollingKeyPairGenerator(X509Certificate2 certificate, X9ECParameters ecParameters)
+            : this(RetrievePrivateKeyBytes(certificate), ecParameters)
+        {
+        }
+
+        private static byte[] RetrievePrivateKeyBytes(X509Certificate2 certificate)
+        {
+            var ecdsaPrivateKey = certificate.GetECDsaPrivateKey();
+            if (ecdsaPrivateKey != null)
+            {
+                return ecdsaPrivateKey.ExportParameters(true).D;
+            }
+
+            var rsaPrivateKey = certificate.GetRSAPrivateKey();
+            if (rsaPrivateKey != null)
+            {
+                // TODO: find a better way to extract bytes for RSA key
+                return Encoding.UTF8.GetBytes(rsaPrivateKey.ToXmlString(true));
+            }
+
+            throw new NotSupportedException($"Unsupported private key for certificate {certificate.Thumbprint}");
+        }
+
+        public BigInteger GetPrivateKey(string keyId)
```

Comment:
The keyId provided here is never validated. At any given point in time, only 1 or 2 keyIds should be valid (current + prev during rollover). The keyId should be validated according to the key pair duration and rollover, similarly to [how it is done in the verification solution](https://github.com/folkehelseinstituttet/Fhi.Smittestopp.Verification/blob/879dd2edd039a7d5d36c425d4b0cebb61a7985ac/Fhi.Smittestopp.Verification.Domain/AnonymousTokens/AnonymousTokenKeyStore.cs#L91). We will then need to sync config for key pair duration between the two services